### ISSUE TITLE
[Question] How to wait for all tasks to finish in WinForms app?

### ISSUE THREAD

**Issue by leoshusar** [microsoft/vs-threading#1161]
I found this library and now I am trying to rewrite my `async void` event handlers with the proper way.

I created my `ThreadHelper` like this:
```cs
public class ThreadHelper : IDisposable
{
    private readonly CancellationTokenSource _disposeCancellationTokenSource = new();

    private readonly JoinableTaskContext _joinableTaskContext;
    private readonly JoinableTaskCollection _joinableTaskCollection;
    private readonly JoinableTaskFactory _joinableTaskFactory;

    public JoinableTaskFactory JoinableTaskFactory => _joinableTaskFactory;
    public JoinableTaskCollection JoinableTaskCollection => _joinableTaskCollection;
    public CancellationToken DisposalToken => _disposeCancellationTokenSource.Token;

    public ThreadHelper()
    {
        _joinableTaskContext = new();
        _joinableTaskCollection = _joinableTaskContext.CreateCollection();
        _joinableTaskFactory = _joinableTaskContext.CreateFactory(_joinableTaskCollection);
    }

    public void Dispose()
    {
        Dispose(true);
        GC.SuppressFinalize(this);
    }

    protected virtual void Dispose(bool disposing)
    {
        if (disposing)
        {
            _disposeCancellationTokenSource.Cancel();

            try
            {
                // Block Dispose until all async work has completed.
#pragma warning disable VSTHRD104 // Offer async methods
                JoinableTaskFactory.Run(JoinableTaskCollection.JoinTillEmptyAsync);
#pragma warning restore VSTHRD104 // Offer async methods
            }
            catch (OperationCanceledException)
            {
                // this exception is expected because we signaled the cancellation token
            }
            catch (AggregateException ex)
            {
                // ignore AggregateException containing only OperationCanceledException
                ex.Handle(inner => inner is OperationCanceledException);
            }
            finally
            {
                _disposeCancellationTokenSource.Dispose();
            }
        }
    }
}
```

I am creating instance of this in my form constructor. Is this okay?

My problem is it deadlocks on the `JoinableTaskFactory.Run(JoinableTaskCollection.JoinTillEmptyAsync);` even if I don't run any task using this factory.

I made simple repro app - I open the app, create `ThreadHelper` instance, do nothing, close the app, call this:
```cs
private void Form1_FormClosing(object sender, FormClosingEventArgs e)
{
    _th.Dispose();
}
```
and deadlock.

Could you please help me with this?

Thank you!

**Comment by AArnott** [comment]
You did a most excellent job, with just the one bug you already found. Here's how you fix it:

```diff
-JoinableTaskFactory.Run(JoinableTaskCollection.JoinTillEmptyAsync);
+_joinableTaskContext.Factory.Run(JoinableTaskCollection.JoinTillEmptyAsync);
```

The reason it deadlocked is you're waiting for a collection to be empty, but you're waiting with a JoinableTask that itself is _in_ that same collection!  You cannot use a JTF instance that adds to the same collection you're waiting on without expecting a deadlock.
With the fix above, you're using the default factory, which _doesn't_ contribute its JoinableTasks to any collection, so the collection has a chance to empty, and you don't deadlock.

**Comment by leoshusar** [comment]
I was on the right track!

While debugging it myself, I found I can look on pending tasks and saw it's in fact waiting for the `JoinTillEmptyAsync`:
![image](https://user-images.githubusercontent.com/13891066/221839780-da5de263-6d2d-4979-b689-89140c6e48b3.png)

but just didn't know why that happens and how to solve it.
And yeah, I saw it in the cookbook, but somehow missed it's two separate instances of JTF.

So your tip worked, thank you!
